File: action_challenge_py/action_challenge_py/move_one_axis_client.py
#!/usr/bin/env python3
from concurrent.futures import Future

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from std_msgs.msg import Empty
from my_robot_interfaces.action import MoveOneAxis
 
class MoveOneAxisClientNode(Node):
    def __init__(self):
        super().__init__("move_one_axis_client")
        self.goal_handle_ : ClientGoalHandle = None
        self.move_one_axis_client_ = ActionClient(self, MoveOneAxis, "move_one_axis")
        self.cancel_subscriber_ = self.create_subscription(
            Empty,
            "cancel_move",
            self.cancel_goal,
            10
        )
        self.get_logger().info("Move One Axis Client has been started.")
        self.get_logger().info("Cancel move via ros2 topic /cancel_move")

    def send_goal(self, target_position, velocity):
        # Wait for the server
        self.move_one_axis_client_.wait_for_server(timeout_sec=10.0)

        # Create a goal
        goal = MoveOneAxis.Goal()
        goal.target_position = target_position
        goal.velocity = velocity

        # Send the goal
        self.get_logger().info("Sending goal")
        self.move_one_axis_client_.send_goal_async(
            goal, feedback_callback=self.goal_feedback_callback
        ).add_done_callback(self.goal_response_callback)

    def cancel_goal(self, msg: Empty):
        self.get_logger().info("Send a cancel request")
        if self.goal_handle_ is not None:
            self.goal_handle_.cancel_goal_async()

    def goal_feedback_callback(self, feedback_msg):
        current_pos = feedback_msg.feedback.current_position
        self.get_logger().info("Current position: %d" %current_pos)

    def goal_response_callback(self, future: Future):
        self.goal_handle_: ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal was accepted by server")
            self.goal_handle_.get_result_async().add_done_callback(
                self.goal_result_callback
            )
        else:
            self.get_logger().error("Goal was rejected by server")

    def goal_result_callback(self, future: Future):
        status = future.result().status
        result = future.result().result
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Success")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Aborted")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn("Canceled")
        self.get_logger().info("Final position: %d" % result.final_position)
        self.get_logger().info(result.message)

 
def main(args=None):
    rclpy.init(args=args)
    node = MoveOneAxisClientNode()
    node.send_goal(100, 2)
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: action_challenge_cpp/src/move_one_axis_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_one_axis.hpp"
#include "std_msgs/msg/empty.hpp"

using MoveOneAxis = my_robot_interfaces::action::MoveOneAxis;
using MoveOneAxisGoalHandle = rclcpp_action::ClientGoalHandle<MoveOneAxis>;

using namespace std::chrono_literals;
using namespace std::placeholders;

 
class MoveOneAxisClientNode : public rclcpp::Node
{
public:
    MoveOneAxisClientNode() : Node("move_one_axis_client")
    {
        move_one_axis_client_ = rclcpp_action::create_client<MoveOneAxis>(this, "move_one_axis");
        cancel_goal_subscriber_ = this->create_subscription<std_msgs::msg::Empty>(
            "cancel_move", 10, std::bind(&MoveOneAxisClientNode::cancel_goal_callback, this, _1)
        );
    }

    void send_goal(int target_position, int velocity)
    {
        // Wait for the Action server
        move_one_axis_client_->wait_for_action_server(10s);

        // Create a goal
        auto goal = MoveOneAxis::Goal();
        goal.target_position = target_position;
        goal.velocity = velocity;

        // Add callbacks
        auto options = rclcpp_action::Client<MoveOneAxis>::SendGoalOptions();
        options.result_callback = std::bind(&MoveOneAxisClientNode::goal_result_callback, 
            this, _1);
        options.goal_response_callback = std::bind(&MoveOneAxisClientNode::goal_response_callback, 
            this, _1);
        options.feedback_callback = std::bind(&MoveOneAxisClientNode::goal_feedback_callback, 
            this, _1, _2);

        // Send the goal
        RCLCPP_INFO(this->get_logger(), "Sending a goal");
        move_one_axis_client_->async_send_goal(goal, options);
    }
private:
    rclcpp_action::Client<MoveOneAxis>::SharedPtr move_one_axis_client_;
    rclcpp::Subscription<std_msgs::msg::Empty>::SharedPtr cancel_goal_subscriber_;
    MoveOneAxisGoalHandle::SharedPtr goal_handle_;

    // Callback to receive the result once the goal is done
    void goal_result_callback(const MoveOneAxisGoalHandle::WrappedResult &result) {
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED) {
            RCLCPP_INFO(this->get_logger(), "Succeeded");
        } else if (status == rclcpp_action::ResultCode::ABORTED) {
            RCLCPP_ERROR(this->get_logger(), "Aborted");
        } else if (status == rclcpp_action::ResultCode::CANCELED) {
            RCLCPP_WARN(this->get_logger(), "Canceled");
        }
        int final_pos = result.result->final_position;
        std::string message = result.result->message;
        RCLCPP_INFO(this->get_logger(), "Final position: %d", final_pos);
        RCLCPP_INFO(this->get_logger(), "Message: %s", message.c_str());
    }

    // Callback to know if the goal was accepted or rejected
    void goal_response_callback(const MoveOneAxisGoalHandle::SharedPtr &goal_handle) {
        if (!goal_handle) {
            RCLCPP_WARN(this->get_logger(), "Goal was rejected by the server");
            return;
        }
        this->goal_handle_ = goal_handle;
        RCLCPP_INFO(this->get_logger(), "Goal was accepted by the server");
    }

    // Callback to receive feedback during goal execution
    void goal_feedback_callback(const MoveOneAxisGoalHandle::SharedPtr &goal_handle,
        const std::shared_ptr<const MoveOneAxis::Feedback> feedback) {
            (void)goal_handle;
            int current_pos = feedback->current_position;
            RCLCPP_INFO(this->get_logger(), "Current position: %d", current_pos);
        }

    void cancel_goal_callback(const std_msgs::msg::Empty::SharedPtr msg) {
        (void)msg;
        RCLCPP_INFO(this->get_logger(), "Canceling the goal");
        move_one_axis_client_->async_cancel_goal(goal_handle_);
    }
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveOneAxisClientNode>();
    node->send_goal(10, 5);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}